src/client/Microsoft.Identity.Client/TokenCache.ITokenCacheInternal.cs

```diff
@@ -264,19 +265,22 @@ private static string GetPreferredUsernameFromIdToken(bool isAdfsAuthority, IdTo
             // no match found after initial filtering
             if (!tokenCacheItems.Any())
             {
-                requestParams.RequestContext.Logger.Info("No matching entry found for user or assertion");
+                logger.Info("No matching entry found for user or assertion");
                 return null;
             }
 
-            requestParams.RequestContext.Logger.Info("Matching entry count - " + tokenCacheItems.Count());
+            if (logger.IsLoggingEnabled(LogLevel.Info))
```

Comment:
Is it possible for the logger to check for this internally?

Reply:
Very good question. It's tricky and I wanted to provide a localized fix without changing "everything". We have a work item to look at this in more detail - https://github.com/AzureAD/microsoft-authentication-library-for-dotnet/issues/1981 which has a prototype for achieving what you are suggesting. 

Essentially in order to delay evaluation of the string to be logged, I was thinking to use a function pointer that generates the string. So instead of giving the logger a string, you give it a function that returns a string. If the log level is not high enough, the logger does not evaluate the function. 

But the problem with this solution is that it needs to be perf tested. Function pointers with closures are complicated and I am not sure that we won't create a different perf problem :). If you want to read more about closures, Jon Skeet has a great article: https://csharpindepth.com/Articles/Closures